nixos/modules/services/web-apps/writefreely.nix

```diff
@@ -0,0 +1,102 @@
+{config, lib, pkgs, ...}:
+
+with lib;
+
+let
+  cfg = config.services.writefreely;
+  configFile = pkgs.writeText "config.ini" (generators.toINI {} cfg.settings);
+  settingsFormat = pkgs.formats.json {};
+in {
+  options.services.writefreely = {
+    enable = mkEnableOption "WriteFreely, a clean, Markdown-based publishing platform made for writers. Write together, and build a community.";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.writefreely;
+      description = ''
+        Overridable attribute of the WriteFreely package to use.
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "writefreely";
+      description = "User account under which writefreely runs.";
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "writefreely";
+      description = "User account under which writefreely runs.";
+    };
+
+    dataDir = mkOption {
+      default = "/var/lib/writefreely";
+      type = types.str;
+      description = "WriteFreely data directory.";
+    };
+
+    settings = lib.mkOption {
+      type = lib.types.submodule {
+        freeformType = settingsFormat.type;
+      };
+      default = {};
+      description = ''
+        Configuration for WriteFreely
+        https://writefreely.org/docs/latest/admin/config
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.writefreely.settings = {
+      server = {
+        templates_parent_dir = lib.mkDefault cfg.package;
+        static_parent_dir = lib.mkDefault cfg.package;
+        pages_parent_dir = lib.mkDefault cfg.package;
+        keys_parent_dir = lib.mkDefault cfg.dataDir;
+      };
+      app = {
+        theme = lib.mkDefault "write";
+      };
+    };
+
+    systemd.services.writefreely = {
+      description = "WriteFreely, a markdown based publishing platform";
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+
+        ExecStartPre = [
+          "${cfg.package}/bin/writefreely -c ${configFile} --gen-keys"
+          "${cfg.package}/bin/writefreely -c ${configFile} --init-db"
+          "${cfg.package}/bin/writefreely -c ${configFile} --migrate"
+        ];
+        ExecStart = "${cfg.package}/bin/writefreely -c ${configFile}";
+        Restart = "on-failure";
+
+        WorkingDirectory = cfg.dataDir;
+        StateDirectory = lib.mkIf
+          (cfg.dataDir == "/var/lib/writefreely")
+          "writefreely";
+        ReadOnlyPaths = [cfg.package];
+        ReadWritePaths = [cfg.dataDir];
+        LimitNOFILE = "1048576";
```

Comment:
Shouldn't this a configurable parameter?

Reply:
what exactly would you want to configure here?

Reply:
Same thing as https://search.nixos.org/options?channel=22.05&show=services.solanum.openFilesLimit&from=0&size=50&sort=relevance&type=packages&query=solanum

Reply:
This can be increased with serviceConfig if really necessary https://search.nixos.org/options?channel=22.05&from=0&size=50&sort=relevance&type=packages&query=serviceConfig